MBS-11558: Show current tracklist data on removal edits

reosarevok opened the pull request:
### Implement MBS-11558

This shows the current tracklist (and a warning) if there are significant changes between the current tracklist for the medium being removed and the tracklist that got stored when entering the removal edit. These are only shown if the edit is open, since otherwise there's probably no real use to them.

Tested with https://musicbrainz.org/edit/78642529 and https://musicbrainz.org/edit/78629796 + a search of open removal edits (but those edits might no longer be open depending on when the testing happens). For testing with local data, just try to enter remove medium edits and then add / remove tracks from the tracklist, add/remove the tracklist completely, and change the tracklist titles or whatnot.

![Screenshot from 2021-12-09 15-41-17](https://user-images.githubusercontent.com/1069224/145407167-ecc606d9-7aca-4d61-a4d0-e6840a0de2f0.png)

yvanzo reviewed (commented):
> Tested with https://musicbrainz.org/edit/78642529 and https://musicbrainz.org/edit/78629796

Tested too but didn’t see any difference, but I guess it is because the edits are closed.

> + a search of open removal edits (but those edits might no longer be open depending on when the testing happens)

The 14 edits currently returned by https://musicbrainz.org/search/edits?auto_edit_filter=&order=desc&negation=0&combinator=and&conditions.0.field=type&conditions.0.operator=%3D&conditions.0.args=53&conditions.1.field=status&conditions.1.operator=%3D&conditions.1.args=1 are not affected by this change.

> For testing with local data, just try to enter remove medium edits and then add / remove tracks from the tracklist, add/remove the tracklist completely, and change the tracklist titles or whatnot.

Would it be worth adding tests?

reosarevok commented:
Added the screenshot of a currently open example to the description. Added a test too, hopefully it looks ok - wasn't quite sure how to test this and where to put it.

reosarevok commented on lib/MusicBrainz/Server/Edit/Medium/Delete.pm:
```diff
@@ -51,7 +51,7 @@ sub foreign_keys
 
     $fk{Medium} = { $self->medium_id => [ 'MediumFormat', 'Release ArtistCredit' ] };
-    $fk{MediumFormat} = { $self->data->{format_id} => [] };
+    $fk{MediumFormat} = { $self->data->{format_id} => [] } if $self->data->{format_id};;
```
It's because format_id is nullable, as defined on the file :) Added a commit message anyway

mwiencek commented on lib/MusicBrainz/Server/Edit/Medium/Delete.pm:
```diff
@@ -51,7 +51,7 @@ sub foreign_keys
 
     $fk{Medium} = { $self->medium_id => [ 'MediumFormat', 'Release ArtistCredit' ] };
-    $fk{MediumFormat} = { $self->data->{format_id} => [] };
+    $fk{MediumFormat} = { $self->data->{format_id} => [] } if $self->data->{format_id};;
```
```suggestion
    $fk{MediumFormat} = { $self->data->{format_id} => [] } if $self->data->{format_id};
```

mwiencek commented on root/edit/details/RemoveMedium.js:
```diff
@@ -9,45 +9,109 @@
+      {showTracklistAndWarning ? (
+        <Warning
+          message={hasLengthChanges
+            ? l(`The number of tracks on the medium being removed
+                 has changed since the removal edit was entered.
+                 Please check the changes and ensure the removal
+                 is still correct.`)
+            : l(`Some track lengths, titles or artists have changed
+                 since the removal edit was entered. Please check the changes
+                 and ensure the removal is still correct.`)
```
It would be cool to show the changes as an EditMedium type diff so that multiple messages aren't needed saying what to look for, but not sure how feasible that is.

reosarevok replied:
It's probably doable, but the calculation of tracklist_changes is not trivial, so I think I'll merge like this and add a ticket for possibly adapting it later.

reosarevok replied:
Added MBS-12236 for that

reosarevok commented on root/edit/details/RemoveMedium.js:
```diff
@@ -9,45 +9,109 @@
-const RemoveMedium = ({edit}: Props): React.Element<'table'> => {
+const areTracksEqual = (a, b) => (
+  a.name === b.name &&
+  reduceArtistCredit(a.artistCredit) === reduceArtistCredit(b.artistCredit) &&
```
I did not! Thanks for the hint.
